### PR TITLE
Always flip choices in BaseStatusType

### PR DESCRIPTION
**Pull request by core23**
I am targeting this branch, because this is a bugfix, but also a soft BC break.

## Changelog

```markdown
### Changed
 - Always flip choices in `BaseStatusType`
```

## Subject

The commit https://github.com/sonata-project/SonataCoreBundle/commit/1eefe7607bbf20ad5ef401febbde6110fb9269f2#diff-c25f4304e73e2ac2efcf019f17558604 introduced a wrong approach which will cause some problems when using symfony 3.

### CODE REVIEW

File: tests/Form/Type/StatusTypeTest.php [hunk 1]
```diff
@@ -60,6 +60,29 @@ public function testGetParent()
...
+    /**
+     * @group legacy
+     */
+    public function testGetDefaultOptionsLegacy()
+    {
+        Choice::$list = [
+            1 => 'salut',
+        ];
+
+        $type = new StatusType('Sonata\CoreBundle\Tests\Form\Type\Choice', 'getList', 'choice_type', true);
```
Comment (OskarStark) [inline tests/Form/Type/StatusTypeTest.php]: Please use `::class constant here

File: tests/Form/Type/StatusTypeTest.php [hunk 2]
```diff
@@ -87,7 +110,7 @@ public function testGetDefaultOptionsWithValidFlip()
             2 => 'toi!',
         ];
 
-        $type = new StatusType('Sonata\CoreBundle\Tests\Form\Type\Choice', 'getList', 'choice_type', true);
+        $type = new StatusType('Sonata\CoreBundle\Tests\Form\Type\Choice', 'getList', 'choice_type');
```
Comment (OskarStark) [inline tests/Form/Type/StatusTypeTest.php]: same here

File: tests/Form/Type/StatusTypeTest.php [hunk 3]
```diff
@@ -109,7 +132,7 @@ public function testGetDefaultOptionsWithValidInvalidFlip()
             2 => 'error',
         ];
 
-        $type = new StatusType('Sonata\CoreBundle\Tests\Form\Type\Choice', 'getList', 'choice_type', true);
+        $type = new StatusType('Sonata\CoreBundle\Tests\Form\Type\Choice', 'getList', 'choice_type');
```
Comment (OskarStark) [inline tests/Form/Type/StatusTypeTest.php]: and here

**Comment by core23** [comment]
Please review again @OskarStark @greg0ire

File: src/Form/Type/BaseStatusType.php [hunk 4]
```diff
@@ -93,6 +103,12 @@ public function configureOptions(OptionsResolver $resolver)
     {
         $choices = call_user_func([$this->class, $this->getter]);
 
+        // choice_as_value options is not needed in SF 3.0+
+        if (method_exists(FormTypeInterface::class, 'setDefaultOptions')) {
```
Comment (greg0ire) [inline src/Form/Type/BaseStatusType.php]: I think @EmmanuelVella has a better way for this, `$resolver->defined('choices_as_value') ` iirc

File: src/Form/Type/BaseStatusType.php [hunk 5]
```diff
@@ -93,6 +103,12 @@ public function configureOptions(OptionsResolver $resolver)
     {
         $choices = call_user_func([$this->class, $this->getter]);
 
+        // choice_as_value options is not needed in SF 3.0+
+        if ($resolver->defined('choices_as_values')) {
```
Comment (EmmanuelVella) [inline src/Form/Type/BaseStatusType.php]: It should be `isDefined ;)`

**Comment by jordisala1991** [comment]
The expected behavior for this type is to pass the choices without the flip applied? This seems a little wrong to use. This would mean that I have to do it like the old way and expect this type to flip the choices, and if I want to do it the new way, then I get a deprecated message.

@core23  why the default value is not false?

**Comment by core23** [comment]
> The expected behavior for this type is to pass the choices without the flip applied?

That's right, but we need this old behavior because of BC reasons. A better way might deprecating this class and create a new one which don't need the flip anymore.

**Pull request merged**